[iOS] Selection in overflow scroller is sometimes offset from selected content while scrolling

whsieh opened the pull request:
#### 1e0a8f21db25c527d049ed36604c90d679f63ad5
<pre>
[iOS] Selection in overflow scroller is sometimes offset from selected content while scrolling
<a href="https://bugs.webkit.org/show_bug.cgi?id=283018">https://bugs.webkit.org/show_bug.cgi?id=283018</a>

Reviewed by Aditya Keerthi.

When `SelectionHonorsOverflowScrolling` is enabled, the selection sometimes (visually) lags behind
the selected content in an overflow scrolling container. This can happen if:

1.  The editor state (which contains selection rects relative to the root view) is computed in the
    web process and sent to the UI process.

2.  While the editor state IPC is being sent via remote layer tree commit, the subscrollable area
    containing selection UI is scrolled (i.e. the content offset changes).

3.  The editor state arrives in the UI process, and we use the stale geometry data from (1) when
    generating selection drawing info.

To fix this, we add logic in the web process to send the scroll position of the enclosing scrollable
area over to the UI process, along with the rest of the editor state&apos;s visual data. In the UI
process, we then use this information to reconcile the enclosing composited scroll view&apos;s scroll
position with the (possibly stale) scroll position when we computed the editor state. If the
difference in scroll offset is more than a tiny delta (1 px), we shift the entire editor state&apos;s
geometry by the necessary amount (mapped to root view coordinates) to ensure that the selection UI
doesn&apos;t get out of sync with its enclosing scroller.

* Source/WebCore/platform/ios/SelectionGeometry.cpp:
(WebCore::SelectionGeometry::move):

Add a helper method to shift the entire `SelectionGeometry` by a given offset.

* Source/WebCore/platform/ios/SelectionGeometry.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebKit/Shared/EditorState.cpp:
(WebKit::EditorState::move):

Add a helper method to shift all owned rects and quads in the `EditorState` by a given offset. Note
that we call `clipOwnedRectExtentsToNumericLimits()` afterwards, to ensure that the editor state
geometry is still robust against underflow / overflow.

* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:

Add `ScrollingNodeID` and `ScrollPosition` to the editor state&apos;s visual data.

* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::reconcileEnclosingScrollViewContentOffset):

See above for more details.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::editorStateChanged):
(WebKit::WebPageProxy::updateEditorState):

Make these take an `EditorState&amp;&amp;` rvalue reference instead of a const reference, so that we can
call into the new `PageClient` hook to reconcile the incoming editor state with subscrollers if
needed.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::reconcileEnclosingScrollViewContentOffset):

Implement the main fix here — check if the enclosing scroller has scrolled since the editor state
was computed, and adjust incoming selection rects.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _reconcileEnclosingScrollViewContentOffset:]):
(-[WKContentView _scrollViewForScrollingNodeID:]):

Pull logic to map a `ScrollingNodeID` to `UIScrollView` out into a separate helper method, so that
we can call it from multiple places.

(-[WKContentView _updateTargetedPreviewScrollViewUsingContainerScrollingNodeID:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::interpretKeyEvent):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::computeEnclosingLayerID const):

Additionally compute the enclosing `ScrollingNodeID` and its corresponding scroll position, and send
it through the editor state update.

Canonical link: <a href="https://commits.webkit.org/286537@main">https://commits.webkit.org/286537@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fa953796e69e461060b5c4fa2c9d460a4b0dd0f5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-intel-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11800 "Built successfully and passed tests") | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-11-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -8457,6 +8457,32 @@ - (void)doAfterEditorStateUpdateAfterFocusingElement:(dispatch_block_t)block
+
+    CGSize contentDelta = [scroller convertRect:CGRectMake(0, 0, scrollDelta.width(), scrollDelta.height()) toView:self].size;
+    contentDelta.width *= scrollDelta.width() >= 0 ? 1 : -1;
+    contentDelta.height *= scrollDelta.height() >= 0 ? 1 : -1;
```
I think we can use `std::copysign` here?

whsieh replied:
Yep, perfect use case for that!

whsieh committed:
[iOS] Selection in overflow scroller is sometimes offset from selected content while scrolling
https://bugs.webkit.org/show_bug.cgi?id=283018

Reviewed by Aditya Keerthi.

When `SelectionHonorsOverflowScrolling` is enabled, the selection sometimes (visually) lags behind
the selected content in an overflow scrolling container. This can happen if:

1.  The editor state (which contains selection rects relative to the root view) is computed in the
    web process and sent to the UI process.

2.  While the editor state IPC is being sent via remote layer tree commit, the subscrollable area
    containing selection UI is scrolled (i.e. the content offset changes).

3.  The editor state arrives in the UI process, and we use the stale geometry data from (1) when
    generating selection drawing info.

To fix this, we add logic in the web process to send the scroll position of the enclosing scrollable
area over to the UI process, along with the rest of the editor state's visual data. In the UI
process, we then use this information to reconcile the enclosing composited scroll view's scroll
position with the (possibly stale) scroll position when we computed the editor state. If the
difference in scroll offset is more than a tiny delta (1 px), we shift the entire editor state's
geometry by the necessary amount (mapped to root view coordinates) to ensure that the selection UI
doesn't get out of sync with its enclosing scroller.

* Source/WebCore/platform/ios/SelectionGeometry.cpp:
(WebCore::SelectionGeometry::move):

Add a helper method to shift the entire `SelectionGeometry` by a given offset.

* Source/WebCore/platform/ios/SelectionGeometry.h:
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebKit/Shared/EditorState.cpp:
(WebKit::EditorState::move):

Add a helper method to shift all owned rects and quads in the `EditorState` by a given offset. Note
that we call `clipOwnedRectExtentsToNumericLimits()` afterwards, to ensure that the editor state
geometry is still robust against underflow / overflow.

* Source/WebKit/Shared/EditorState.h:
* Source/WebKit/Shared/EditorState.serialization.in:

Add `ScrollingNodeID` and `ScrollPosition` to the editor state's visual data.

* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::reconcileEnclosingScrollViewContentOffset):

See above for more details.

* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTreeTransaction):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::editorStateChanged):
(WebKit::WebPageProxy::updateEditorState):

Make these take an `EditorState&&` rvalue reference instead of a const reference, so that we can
call into the new `PageClient` hook to reconcile the incoming editor state with subscrollers if
needed.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::reconcileEnclosingScrollViewContentOffset):

Implement the main fix here — check if the enclosing scroller has scrolled since the editor state
was computed, and adjust incoming selection rects.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _reconcileEnclosingScrollViewContentOffset:]):
(-[WKContentView _scrollViewForScrollingNodeID:]):

Pull logic to map a `ScrollingNodeID` to `UIScrollView` out into a separate helper method, so that
we can call it from multiple places.

(-[WKContentView _updateTargetedPreviewScrollViewUsingContainerScrollingNodeID:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::interpretKeyEvent):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::computeEnclosingLayerID const):

Additionally compute the enclosing `ScrollingNodeID` and its corresponding scroll position, and send
it through the editor state update.

Canonical link: https://commits.webkit.org/286537@main
